### PR TITLE
Create experimental postgis driver

### PR DESCRIPTION
**Pull request by SpacemanPaul**
### Reason for this pull request

Create a clone of the existing postgres index driver as the postgis index driver, and flag as experimental.

First step on [the ODCv2 Roadmap](https://github.com/opendatacube/datacube-core/wiki/ODC-v2-Roadmap)

Includes:

- changes to integration test fixture parametisation so that tests can be run for both index drivers in parallel.
- documentation, typehint and test cleanups identified by @omad in his review for PR #1247

 - [*] Tests added / passed
 - [*] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: docs/installation/extending-odc.rst [hunk 1]
```diff
@@ -254,6 +257,17 @@ The memory index driver may be useful:
 3. as an example/template for developing other index drivers.
 
 
+Postgis Implementation
```
Comment (omad) [inline docs/installation/extending-odc.rst]: PostGIS

**Review by omad — APPROVED** [review]
This is just a copy/paste search/replace of the postgres driver? Plus a couple more bits and pieces.

File: docs/installation/extending-odc.rst (hunk 1 above)
Comment (SpacemanPaul) [reply docs/installation/extending-odc.rst]: Yes, "other bits and pieces" being mostly a slight refactor of the integration test fixtures to allow tests to be run against both drivers.

(which works for me beautifully locally, but is failing on github for reasons I have yet to figure out)

**Comment by SpacemanPaul** [comment]
Tests are failing because:

- The GHA docker-test-runner.yml workflow both BUILDS and RUNS the integration tests. If it's successful, AND is running from a checkout of the develop, it pushes the image that it built to Docker Hub as opendatacube/datacube-test:latest.
- The GHA main.yml, Doesn't build a docker image, it DOWNLOADS opendatacube/datacube-test:latest, and the runs tests using the current version of the code mounted inside that image.

Because the image isn't rebuilt, it's not using any of my updates to the with_bootstrap script, hence all the experimental driver tests fail.

**Pull request merged**